libraries/Microsoft.Bot.Builder.Expressions/BuiltInFunctions.cs

```diff
@@ -1229,623 +1230,1733 @@ private static (object, string) ParseISOTimestamp(string timeStamp, Func<DateTim
+                {
+                    result = destanationTS.ToString(format);
+                }
+                catch
+                {
+                    error = "illegal format representation";
                 }
             }
 
             return (result, error);
         }
 
-        private static (object, string) SubArray(Expression expression, object state)
+        private static (object, string) StartOfDay(string timestamp, string format = "o")
         {
-            object result = null;
-            string error;
-            object arr;
-            (arr, error) = expression.Children[0].TryEvaluate(state);
+            string result = null;
+            dynamic ts = null;
+            string error = null;
+            try
+            {
+                ts = DateTime.Parse(timestamp).ToUniversalTime();
+            }
+            catch
+            {
+                error = "illegal timestamp format";
+            }
 
             if (error == null)
             {
-                if (TryParseList(arr, out var list))
+                var startOfDay = ts.Date;
+                try
                 {
-                    var startExpr = expression.Children[1];
-                    object startObj;
-                    (startObj, error) = startExpr.TryEvaluate(state);
-                    var start = 0;
-                    if (error == null)
-                    {
-                        if (startObj == null || !startObj.IsInteger())
-                        {
-                            error = $"{startExpr} is not an integer.";
-                        }
-                        else
-                        {
-                            start = (int)startObj;
+                    result = startOfDay.ToString(format);
+                }
+                catch
+                {
+                    error = "illegal format representation";
+                }
+            }
+
+            return (result, error);
+        }
+
+        private static (object, string) StartOfHour(string timestamp, string format = "o")
+        {
+            string result = null;
+            dynamic ts = null;
+            string error = null;
+            try
+            {
+                ts = DateTime.Parse(timestamp).ToUniversalTime();
+            }
+            catch
+            {
+                error = "illegal timestamp format";
+            }
+
+            if (error == null)
+            {
+                var startOfDay = ts.Date;
+                var hours = ts.Hour;
+                var startOfHour = startOfDay.AddHours(hours);
+                try
+                {
+                    result = startOfHour.ToString(format);
+                }
+                catch
+                {
+                    error = "illegal format representation";
+                }
+            }
+
+            return (result, error);
+        }
+
+        private static (object, string) StartOfMonth(string timestamp, string format = "o")
```

Comment:
Could 'startOfxxx' extracted into one function? use TimeUnit string to seperate them